boards: Add support for Adafruit Feather M0 board with LoRa radio

migdard opened the pull request:
This adds support for the Cortex M0-based boards from Adafruit that include the Semtech LoRa radio chip.

Note that this is mostly based on the adafruit_feather_m0_basic_proto board.

Signed-off-by: Miguel Dardenne <miguel.dardenne@gmail.com>

pdgendt commented:
Please read the [Contribution Guidelines](https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html) to clean up the PR commits.

The LoRa radio is a simple addition to the already existing basic Adafruit Feather M0 board, please create a board variant instead of duplicating the existing one entirely.

migdard commented:
> Please read the [Contribution Guidelines](https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html) to clean up the PR commits.
> 
> The LoRa radio is a simple addition to the already existing basic Adafruit Feather M0 board, please create a board variant instead of duplicating the existing one entirely.

I will cleanup the PR accordingly, thanks for the link.

Regarding creating a board variant instead of duplicating (most of) the existing one, I considered this but only could find the concept of  [board revision](https://docs.zephyrproject.org/3.1.0/hardware/porting/board_porting.html#multiple-board-revisions), which as I understand follows a MAJOR.MINOR.PATCH numbered versioning scheme.

I assume following the [shield approach](https://docs.zephyrproject.org/3.1.0/hardware/porting/shields.html) is closer to the goal of minimising duplicated code and even if, technically, the Lora chip is not soldered or attached by the user (as naming it a shield would imply). ?


Thanks again

pdgendt commented:
> I assume following the [shield approach](https://docs.zephyrproject.org/3.1.0/hardware/porting/shields.html) is closer to the goal of minimising duplicated code and even if, technically, the Lora chip is not soldered or attached by the user (as naming it a shield would imply). ?

I think it would be best to simply add your board next to the existing one (maybe even rename `adafruit_feather_m0_basic_proto` to `adafruit_feather_m0`). And have both `basic_proto` and `lora`.
For an example see [`mimxrt1060_evk`](https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/mimxrt1060_evk) where a hyperflash variant exists.

migdard commented:
> I think it would be best to simply add your board next to the existing one, in the same directory (maybe even rename `adafruit_feather_m0_basic_proto` to `adafruit_feather_m0`). And have both `basic_proto` and `lora`. For an example see [`mimxrt1060_evk`](https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/mimxrt1060_evk) where a hyperflash variant exists.

I think it looks better now. There are still 2 failed unit tests but I believe they were there before. A vanilla version from main does not compile with `CONFIG_UART_ASYNC_API=y` so I am pretty sure async serial is broken for that board since a while. Given I will need serial to work for my project, I will see what I can do (in another potential PR).

pdgendt commented on boards/arm/adafruit_feather_m0/Kconfig.board:
```diff
@@ -1,8 +1,13 @@
 	depends on SOC_PART_NUMBER_SAMD21G18A
+
+config BOARD_ADAFRUIT_FEATHER_M0_LORA
+    bool "Adafruit Feather M0 LoRa"
+    depends on SOC_PART_NUMBER_SAMD21G18A
```
Fix indentation, use tabs

migdard replied:
Done

pdgendt commented on boards/arm/adafruit_feather_m0/adafruit_feather_m0_lora.dts:
```diff
@@ -0,0 +1,30 @@
+#include "adafruit_feather_m0_basic_proto.dts"
+
+&sercom4 {
+	status = "okay";
+	compatible = "atmel,sam0-spi";
+	#address-cells = <1>;
+	#size-cells = <0>;
+	dipo = <0>;
+	dopo = <1>;
+
+	pinctrl-0 = <&sercom4_spi_default>;
+	pinctrl-names = "default";
```
No need to duplicate this.

migdard replied:
Done

pdgendt commented on boards/arm/adafruit_feather_m0/adafruit_feather_m0_lora.dts:
```diff
@@ -0,0 +1,30 @@
+	pinctrl-0 = <&sercom4_spi_default>;
+	pinctrl-names = "default";
+
+    cs-gpios = < &porta 6 GPIO_ACTIVE_LOW>;
+    lora:radio@0 {
+        reg = <0>;
+        compatible = "semtech,sx1276";
+        reset-gpios = <&porta 8 GPIO_ACTIVE_LOW>;
+        dio-gpios = <&porta 9  (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
+        spi-max-frequency = <10000000>;
+        power-amplifier-output = "pa-boost";
+    };
```
Fix indentation

pdgendt reviewed (changes requested):
Review the [commit guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and fix:

- No merge commits, `rebase` onto `main`
- No fixup commits of something that is part of the PR, squash those commits together

pdgendt commented on boards/arm/adafruit_feather_m0/adafruit_feather_m0_lora.dts:
```diff
@@ -0,0 +1,20 @@
+/dts-v1/;
+#include "adafruit_feather_m0_basic_proto.dts"
+
+&sercom4 {
+	cs-gpios = < &porta 6 GPIO_ACTIVE_LOW>;
+	lora:radio@0 {
+		reg = <0>;
+		compatible = "semtech,sx1276";
+		reset-gpios = <&porta 8 GPIO_ACTIVE_LOW>;
+		dio-gpios = <&porta 9  (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
+		spi-max-frequency = <10000000>;
+		power-amplifier-output = "pa-boost";
+	};
+};
```
Whitespace nits:

```suggestion
&sercom4 {
	cs-gpios = <&porta 6 GPIO_ACTIVE_LOW>;
	lora: radio@0 {
		reg = <0>;
		compatible = "semtech,sx1276";
		reset-gpios = <&porta 8 GPIO_ACTIVE_LOW>;
		dio-gpios = <&porta 9 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
		spi-max-frequency = <10000000>;
		power-amplifier-output = "pa-boost";
	};
};
```

pdgendt reviewed (changes requested):
Looks good, but the board image file size is 4x larger than the basic proto variant. Because of growing concerns for repository sizes, we should keep these images small.
Resize to a max width of 640px.

pdgendt commented:
I think the CI build failures are unrelated to the changes, but will need to be resolved as well. Maybe @nandojve can help out here?

migdard commented:
> I think the CI build failures are unrelated to the changes, but will need to be resolved as well. Maybe @nandojve can help out here?

Agree, and I believe I have narrowed it down already.
Whenever `CONFIG_UART_ASYNC_API` is set, one needs to have some properties (dmas, dma-names) defined for the  sam0_uart driver. If they are not (which is the case on this board), compilation fails.


Should we provide working default values for these in the devicetree ?

migdard commented:
> > I think the CI build failures are unrelated to the changes, but will need to be resolved as well. Maybe @nandojve can help out here?
> 
> Agree, and I believe I have narrowed it down already. Whenever `CONFIG_UART_ASYNC_API` is set, one needs to have some properties (dmas, dma-names) defined for the sam0_uart driver. If they are not (which is the case on this board), compilation fails.
> 
> Should we provide working default values for these in the devicetree ?

As mentioned in https://github.com/zephyrproject-rtos/zephyr/discussions/50838 I propose to exclude this variant from the serial async API testing as well (the basic_proto variant was already excluded).

migdard commented on boards/arm/adafruit_feather_m0/adafruit_feather_m0_lora.yaml:
```diff
@@ -0,0 +1,10 @@
+type: mcu
+arch: arm
+ram: 32
+flash: 256
```
Either way I think this is no different than was already the case for the basic_proto variant (see https://github.com/zephyrproject-rtos/zephyr/blob/1d879b93c59e8f1f48d86db5794600769c86facb/boards/arm/adafruit_feather_m0_basic_proto/adafruit_feather_m0_basic_proto.yaml ) .

Should we open a separate issue to fix for both variants ?

nandojve replied:
Since you are renaming here I think you can just create another commit after to fix it.
If you prefer, you could split into 3 commits to make clearly:
* preparation (the renaming to track history)
* your changes
* the fix

henrikbrixandersen reviewed (commented):
> The LoRa radio is a simple addition to the already existing basic Adafruit Feather M0 board, please create a board variant instead of duplicating the existing one entirely.

While this may be true, the vendor (Adafruit) does not list these boards as being variant of each other. The board variant approach pretty much copies every file anyways, so I don't believe we should add this as a board variant, but as a separate board (as this is what the vendor lists it as).

henrikbrixandersen commented:
> Not sure if I misused the terminology, but it is a different board you can build for. Or do you mean to have both top-level board folders `adafruit_feather_m0_basic_proto` and `adafruit_feather_m0_lora`?

Yes.

> If, in best case, all feather m0 boards would be added to zephyr, this would mean we would have 11 boards with almost the same devicetree.

But by grouping them and reusing the basic DTS you are seeing patterns/reusability where the vendor maybe did not intend for a pattern. If the vendor had listed them as a board family with different variants, I would be in favour of us representing them like that as well. But the vendor lists the entire Feather series as a board family with each board on its own.

pdgendt commented:
> If the vendor had listed them as a board family with different variants, I would be in favour of us representing them like that as well. But the vendor lists the entire Feather series as a board family with each board on its own.

How do we expect the vendor to make this explicit? Currently on the product page for the LoRa board, it states:

> Here's some handy specs! **Like all Feather M0**'s you get:

And a bit later:

> This Feather M0 LoRa Radio uses the extra space left over to add an RFM9x LoRa 868/915 MHz radio module.

migdard commented:
> But by grouping them and reusing the basic DTS you are seeing patterns/reusability where the vendor maybe did not intend for a pattern. If the vendor had listed them as a board family with different variants, I would be in favour of us representing them like that as well. But the vendor lists the entire Feather series as a board family with each board on its own.

So.. we have two conflicting opinions here. I personally think that both views are valid, but this particular point wins for me. Why would both boards share sercom0 configured as UART, sercom3 as I2C, sercom4 as SPI, etc. while in principle there should be more than one way to configure each one ? I do not even know why they are configured like this by default in the basic proto dts... demo/sample? testability?

pdgendt commented:
> So.. we have two conflicting opinions here. 

Yes, however I'm fine with either outcome. I'll abstain from approving to have at least two other reviewers on board (pun intended).

> Why would both boards share sercom0 configured as UART, sercom3 as I2C, sercom4 as SPI, etc. while in principle there should be more than one way to configure each one ? I do not even know why they are configured like this by default in the basic proto dts... demo/sample? testability?

I think it is helpful for tests and samples which can be mapped using aliases to some instance of a device (UART, I2C, SPI, ...) without the need to add an overlay to each test.

migdard commented:
> > Not sure if I misused the terminology, but it is a different board you can build for. Or do you mean to have both top-level board folders `adafruit_feather_m0_basic_proto` and `adafruit_feather_m0_lora`?
> 
> Yes.


This is addressed in latest update. @henrikbrixandersen, if you would kindly review again.

Thanks

migdard commented:
> Review the [commit guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and fix:
> 
> * No merge commits, `rebase` onto `main`
> * No fixup commits of something that is part of the PR, squash those commits together

Addressed as well.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_basic_proto/Kconfig.defconfig:
```diff
@@ -1,4 +1,4 @@
-# Adafruit Feather M0 Basic Proto board configuration
+# Adafruit Feather M0 boards configuration
```
Skip this, please.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_basic_proto/adafruit_feather_m0_basic_proto.yaml:
```diff
@@ -3,7 +3,7 @@ name: Adafruit Feather M0 Basic Proto
 arch: arm
 ram: 32
-flash: 256
+flash: 232
```
How come you are reducing the amount of flash reported?

migdard replied:
Not finding the comment (deleted?), but someone had mentioned that the size of the code partition had to be taken into account to prevent some tests from failing. Code partition is declared in the device tree with a size of 0x3A000 = 232kB. This is where the value comes from.

henrikbrixandersen reviewed (changes requested):
The Git commit log needs to be revised.

migdard commented:
> The Git commit log needs to be revised.

I skimmed through the contribution guidelines and hopefully now it is better.
Please let me know how, if not.

pdgendt commented on boards/arm/adafruit_feather_m0_basic_proto/adafruit_feather_m0_basic_proto.yaml:
```diff
@@ -3,7 +3,7 @@ name: Adafruit Feather M0 Basic Proto
 arch: arm
 ram: 32
-flash: 256
+flash: 232
```
This belongs in a separate commit.

henrikbrixandersen replied:
Or even a separate PR.

migdard replied:
Separate PR https://github.com/zephyrproject-rtos/zephyr/pull/51618 opened for this change.

migdard committed:
boards: Add support for Adafruit Feather M0 with LoRa radio

Add support for the Feather M0 LoRa board, including support of its
Semtech Lora radio. Tested the radio using a pair of boards.

Signed-off-by: Miguel Dardenne <miguel.dardenne@gmail.com>

pdgendt commented on boards/arm/adafruit_feather_m0_lora/adafruit_feather_m0_lora.dts:
```diff
@@ -0,0 +1,125 @@
+	pinctrl-names = "default";
+};
+
+&sercom4 {
+	cs-gpios = <&porta 6 GPIO_ACTIVE_LOW>;
+	lora: radio@0 {
+		reg = <0>;
+		compatible = "semtech,sx1276";
+		reset-gpios = <&porta 8 GPIO_ACTIVE_LOW>;
+		dio-gpios = <&porta 9 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
+		spi-max-frequency = <10000000>;
+		power-amplifier-output = "pa-boost";
+	};
+};
```
I'd prefer a single `&sercom4` section in this file.

migdard replied:
Me too actually :) Done.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/adafruit_feather_m0_lora_defconfig:
```diff
@@ -0,0 +1,13 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_GPIO=y
```
Is `CONFIG_GPIO=y` needed for basic operation? If not, it should not be enabled on the board level.

migdard replied:
It can probably be removed. I based this on the equivalent in the Basic Proto defconfig ( https://github.com/zephyrproject-rtos/zephyr/blob/a61ca79041a0b8679a77a78228fff7698e2be1d2/boards/arm/adafruit_feather_m0_basic_proto/adafruit_feather_m0_basic_proto_defconfig#L10 ) .

If confirmed as not needed, should I file another PR for fixing the Basic Proto too ?

migdard replied:
Included in https://github.com/zephyrproject-rtos/zephyr/pull/51920

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
++-----------+------------+------------------------------------------+
+| USB       | on-chip    | USB device                               |
++-----------+------------+------------------------------------------+
+| RADIO     | SPI        | LoRa radio (on Lora variant only)        |
```
```suggestion
| RADIO     | SPI        | LoRa radio         |
```

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+===================
+
+The `Adafruit Feather M0 with LoRa radio module Learn site`_ has detailed
+information about the Basic Proto board variant including `pinouts`_
```
Forgot to update documentation here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+===========
+
+The SAMD21 MCU has 6 SERCOM based USARTs.  On the Adafruit Feather M0
+Basic Proto, SERCOM0 is the Zephyr console and is available on pins 0
```
and here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+========
+
+The SAMD21 MCU has 6 SERCOM based USARTs.  On the Adafruit Feather M0
+Basic Proto, SERCOM3 is available on pin 20 (SDA) and pin 21 (SCL).
```
and here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+========
+
+The SAMD21 MCU has 6 SERCOM based SPIs.  On the Adafruit Feather M0
+Basic Proto, SERCOM4 is available on pin 22 (MISO), pin 23 (MOSI), and
```
and here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+
+LoRa Radio
+==========
+The Semtech SX127x radio chip on the LoRa variant is attached to the
```
No need to mention a variant here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+Programming and Debugging
+*************************
+
+The Adafruit Feather M0 Basic Proto ships with a BOSSA compatible
```
and here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+      :goals: build
+      :compact:
+
+#. Connect the Adafruit Feather M0 Basic Proto to your host computer
```
and here.

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+      :goals: flash
+      :compact:
+
+   You should see "Hello World! arm" in your terminal.
```
arm?

migdard replied:
Yes, see https://github.com/zephyrproject-rtos/zephyr/blob/a61ca79041a0b8679a77a78228fff7698e2be1d2/samples/hello_world/src/main.c#L11

migdard replied:
Right, should be changed.
Same issue in upstream board https://github.com/zephyrproject-rtos/zephyr/blame/main/boards/arm/adafruit_feather_m0_basic_proto/doc/index.rst#L153

migdard replied:
And actually a generic issue since https://github.com/zephyrproject-rtos/zephyr/commit/e36c7f5b2bff88b2bc4302325a1e39c1265cce5b

henrikbrixandersen commented on boards/arm/adafruit_feather_m0_lora/doc/index.rst:
```diff
@@ -0,0 +1,180 @@
+
+.. target-notes::
+
+.. _Adafruit Feather M0 Basic Proto Learn site:
+    https://learn.adafruit.com/adafruit-feather-m0-basic-proto/
```
Needs updating.

henrikbrixandersen reviewed (changes requested):
You seem to have reused a lot of copyright notices in the committed files. This may be alright if you have based the files on someone else's work.

nandojve commented on boards/arm/adafruit_feather_m0_lora/adafruit_feather_m0_lora.dts:
```diff
@@ -0,0 +1,123 @@
+		 */
+		storage_partition: partition@3c000 {
+			label = "storage";
+			reg = <0x0003c000 0x00004000>;
```
Use DT_SIZE_foo

migdard replied:
Done and included in https://github.com/zephyrproject-rtos/zephyr/pull/51920

lbdroid commented:
I'm trying to build lora sample to test on this board and getting a build error;

```$  west build -p auto -b adafruit_feather_m0_lora samples/drivers/lora/send/
[13/100] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj 
ccache /usr/bin/arm-none-eabi-gcc -DKERNEL -D__PROGRAM_START -D__ZEPHYR__=1 -I/mnt/data/temp/zephyr/zephyr/include -I/mnt/data/temp/zephyr/zephyr/build/zephyr/include/generated -I/mnt/data/temp/zephyr/zephyr/soc/arm/atmel_sam0/samd21 -I/mnt/data/temp/zephyr/zephyr/drivers -I/mnt/data/temp/zephyr/zephyr/soc/arm/atmel_sam0/common/. -I/mnt/data/temp/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/mnt/data/temp/zephyr/modules/hal/atmel/include -I/mnt/data/temp/zephyr/modules/hal/atmel/asf/sam0/include/samd21 -isystem /mnt/data/temp/zephyr/zephyr/lib/libc/minimal/include -isystem /usr/lib/gcc/arm-none-eabi/12.2.0/include -isystem /usr/lib/gcc/arm-none-eabi/12.2.0/include-fixed -fno-strict-aliasing -Os -imacros /mnt/data/temp/zephyr/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m0plus -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/usr/arm-none-eabi -imacros /mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send=CMAKE_SOURCE_DIR -fmacro-prefix-map=/mnt/data/temp/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/mnt/data/temp/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c
In file included from /mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain.h:50,
                 from /mnt/data/temp/zephyr/zephyr/include/zephyr/devicetree/zephyr.h:15,
                 from /mnt/data/temp/zephyr/zephyr/include/zephyr/devicetree.h:3931,
                 from /mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:12,
                 from /mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c:7:
/mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain/gcc.h:77:36: error: static assertion failed: "No default LoRa radio specified in DT"
   77 | #define BUILD_ASSERT(EXPR, MSG...) _Static_assert(EXPR, "" MSG)
      |                                    ^~~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c:14:1: note: in expansion of macro 'BUILD_ASSERT'
   14 | BUILD_ASSERT(DT_NODE_HAS_STATUS(DEFAULT_RADIO_NODE, okay),
      | ^~~~~~~~~~~~
In file included from /mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain/gcc.h:88:
/mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c: In function 'main':
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_ALIAS_lora_ORD' undeclared (first use in this function)
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain/common.h:132:26: note: in definition of macro '_DO_CONCAT'
  132 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:83:33: note: in expansion of macro '_CONCAT'
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:209:37: note: in expansion of macro 'DEVICE_NAME_GET'
  209 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:226:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  226 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c:27:47: note: in expansion of macro 'DEVICE_DT_GET'
   27 |         const struct device *const lora_dev = DEVICE_DT_GET(DEFAULT_RADIO_NODE);
      |                                               ^~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:83:41: note: each undeclared identifier is reported only once for each function it appears in
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/toolchain/common.h:132:26: note: in definition of macro '_DO_CONCAT'
  132 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:83:33: note: in expansion of macro '_CONCAT'
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:209:37: note: in expansion of macro 'DEVICE_NAME_GET'
  209 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/include/zephyr/device.h:226:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  226 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/mnt/data/temp/zephyr/zephyr/samples/drivers/lora/send/src/main.c:27:47: note: in expansion of macro 'DEVICE_DT_GET'
   27 |         const struct device *const lora_dev = DEVICE_DT_GET(DEFAULT_RADIO_NODE);
      |                                               ^~~~~~~~~~~~~
[22/100] Building C object zephyr/lib/libc/minimal/CMakeFiles/lib__libc__minimal.dir/source/stdlib/malloc.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /mnt/data/temp/zephyr/zephyr/build
```

Any hint on how to make it work?

fabiobaltieri commented:
@lbdroid would you be able to submit a pull request with the fix? This is the [contribution guideline documentation](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-guidelines).

migdard commented:
#worksforme

migdard commented:
Adding 
lora0 = &lora;
to adafruit_feather_m0_lora.dts in the aliases section fixes it for me
